### DISCUSSION TITLE
VGA Double Scan and How It Handles Horizontal Resolution

### DISCUSSION THREAD

**Discussion by anikom15** [86Box/86Box#3323] [General]
I believe the current behavior of 86Box's VGA emulation is incorrect. The line-doubling of the vertical resolution is well-known, so that a 200 line mode will appear as 400 lines. This seems to be done properly as of now. The issue is  the handling of horizontal resolution. Right now, it appears that the horizontal resolution is also doubled when 320x200 mode is used. I'm not sure how it handles 640x200 as that is rare and I don't have any software to test with that mode right now.

I believe this behavior is incorrect. There was absolutely no need for horizontal resolution scaling in that time period, because monitors didn't care about the horizontal resolution. To the monitor, it simply saw lines as lines. Furthermore, the added complexity of having logic to double pixels when low horizontal resolution modes are used for something that isn't needed by the hardware is rather uncharacteristic of the design sensibilities of IBM. So, I am strongly skeptical that the emulation is correct.

Checking in the IBM VGA/XGA technical reference, there is some evidence that not doubling the horizontal resolution is the correct behavior:

'In the 200-scan-line modes, the data for each scan line is scanned
twice. This double scanning allows the 200-scan-line image to be
displayed in 400 scan lines.' p. 2-12

In the 640x200 mode:
'Addressing, scan-line mapping, and data format are the same as
the 640 x 200 PEL black and white mode of the IBM Color/Graphics
Monitor Adapter. The display buffer is configured at hex B8000.
Bit image data is stored in memory map 0 and comprises a single
bit plane (C0). Features of this mode are:
...
Double scanned to display as 400 rows
...'

Since obviously the 640 horizontal pixels could never have been doubled to 1280 on actual hardware, it's apparent that the term 'double scan' in the reference is solely referring to the vertical lines, and not to horizontal pixels. The duplication of pixels is never mentioned in the reference document.

The implication for this is that this behavior has been wrong on PCem-based emulators for years. A bug report should be filed and the behavior should be fixed as soon as possible.

**Comment by anikom15** [comment]
I have rewritten the CGA driver to output at its proper resolution, depending on the mode this can be either 320x200 or 640x200. Before it was 'buffering' the 320 modes to 640 pixels and buffering all modes to 400 lines.

You can see the changes here: https://github.com/anikom15/86Box/tree/cga

I intend to open a pull request, but I'd like to give the devs an opportunity to comment. I see this as a trial for implementing this behavior in the rest of the graphics drivers.

The CGA driver also needs some additional work. The overscan is wrong, and the code is somewhat difficult to read. Though I will prioritize the res-setting behaviors of the other modes so long as they can easily be tested and their behavior is at least somewhat -sufficiently documented.

The correct resolution setting is important for shaders. You can see the effect of the same shader on the old behavior vs. the new behavior:
**Old - Incorrect**
<img width="1440" alt="cga1" src="https://user-images.githubusercontent.com/3195646/237027008-4c37969e-cf68-417e-beca-50f14d815235.png">

**New - Correct**
<img width="1440" alt="cga2" src="https://user-images.githubusercontent.com/3195646/237027022-7f36002c-804f-4ab6-8984-235faec3e887.png">

**Comment by OBattler** [comment]
The thing is that CGA does double the resolution. This is a photo of the IBM CGA text mode: 
![imagem](https://github.com/86Box/86Box/assets/10365853/1a18a9a2-e674-439c-8b44-3b8069754555)
It clearly shows that every character is doubled both vertically and horizontally. This even affects composite mode - it's why the 40x25 mode (doubled both horizontally and vertically) is more legible on composite than 80x25 (which is doubled only vertically).

**Comment by OBattler** [comment]
Here is another photo: ![imagem](https://github.com/86Box/86Box/assets/10365853/f7d9f217-a969-4dc0-a39e-07f2eaa206f5) .

**Reply by anikom15** [reply]
Thanks. The images are a bit too low res for me to distinguish clearly. Perhaps they got compressed when you uploaded them. That said, I was able to find a video of someone comparing how CGA graphics appear on a CGA (15 kHz monitor) vs. VGA (31 kHz monitor) card:
https://www.youtube.com/watch?v=hQBoAyRo624

I'm not sure I understand what you mean by the composite output being 480i. The composite output of the card is not interlaced, and has the same vertical resolution of the RGBi mode. Both modes are output using '240p', which was how computer systems at the time displayed progressive scan signals. This is accomplished by shifting the vsync signal to 'trick' the monitor into displaying odd and even scanlines in the same place. A 400 line progressive signal cannot be synced on an NTSC monitor, as that would require a 31 kHz horizontal frequency. NTSC uses a 15 kHz horizontal frequency. You can find more information on how old systems used progressive scan here:
https://en.wikipedia.org/wiki/Low-definition_television
http://scanlines.hazard-city.de/

I'm not sure how 40 column mode being more legible than 80 column mode indicates horizontal pixel doubling. The Apple II, Atari 8-bit computers, and Commodore 64 all had 40 column text that was legible without needing to double the pixels. If anything, pixel doubling could reduce legibility, because high-frequency blips caused by a pixel doubler would appear as an artifact color on a composite monitor.

It appears someone is working on redrawing the CGA schematics:
https://github.com/hkzlab/CGA_Schematics

**Reply by OBattler** [reply]
The composite output of the card has to be interlaced because it's standard NTSC, as it's meant to be plugged into a regular TV. Hence, it's 480i. NTSC is by definition interlaced, as are PAL and SÉCAM. NTSC is 60 fields per second, each fieild being half a frame, so one second the even frame is sent and another second the odd frame is sent. A standard NTSC TV, especially one from the 80's, isn't going to be able to sync into a progressive signal of any kind.

I believe the CGA outputs 30 progressive frames per second which then the composite output converts to 60 interlaced fields per second, as required by the NTSC standard. The NTSC standard is interlaced, with 525 lines per frame, of which 480 are visible.

Also, the horizontal resolution is defined as 40 or 80 character clocks, how many pixels per character clock are there, is unclear for the CGA. The VGA, on the other hand, explicitly has a "low resolution" bit that horizontally doubles every pixel, it's used for both the 40-column text modes, as well as for stuff like 320x200. This is in addition to a vertical line doubler bit.

Not sure about the VGA, but the IBM CGA technical reference hints at this being achievable by reducing the frequency to 7 MHz, ie. basically halving the frequency:
> The adapter operates in noninterlace mode at either 7 or 14 MHz, depending on the mode of the operation selected.
Now, yes, it says noninterlace, but it *has* to be converted to interlaced on composite out, per the NTSC standard, as IBM says the CGA is intended to be compatible with home TV's. In addition, it has an explicit bit to enable interlaced mode as well.

Also, I assume that at 7 MHz, all the timings are halved, which is exactly what is needed for the modes with lower horizontal resolutions, so that the time taken to beam those 40 characters is the same as it would normally take for 80 characters.

> This is accomplished by shifting the vsync signal to 'trick' the monitor into displaying odd and even scanlines in the same place.

While a this can be displayed on a regular TV, Wikipedia says this would result in the scanline pattern. Now, this is a capture of actual CGA composite out: 
![imagem](https://github.com/86Box/86Box/assets/10365853/44b8a8a5-78d2-48f6-a4b2-a9e63774978b) . I believe this comes from Reenigne or one of the other guys who worked on 8088 MPH. I don't see such a scanline pattern.

**Reply by OBattler** [reply]
Well, I'll be damned, I just found this higher resolution photo of a CGA monitor: 
![imagem](https://github.com/86Box/86Box/assets/10365853/1217373e-65f6-4f76-adc0-a0f4b5bdedcd) and it indeed shows the scanline pattern. Now I just need a photo of a real TV showing CGA composite output.

**Reply by OBattler** [reply]
Found one for cmposite as well now: 
![imagem](https://github.com/86Box/86Box/assets/10365853/41e64136-0ff6-4907-b726-6c2595ab4487) and I also see the same scanline pattern. So you were right.

**Reply by OBattler** [reply]
However, this is about the vertical resolution. For horizontal, what happens is that the beam changes intensity every X seconds, and then it depends on the size of the grid on the screen.

**Reply by anikom15** [reply]
Looking at the reference document, there are references to 'picture elements' or PELs, which is an old-fashioned way of referring to a pixel. There are descriptions of the graphics modes:

### Low-Resolution Color/Graphics Mode
The low-resolution mode supports home televisions or color
monitors. This mode, not supported in ROM, has the following
features:
* Contains a maximum of 160 PELs by 100 rows, with each
PEL being 2-high by 2-wide.
...

The description '2-high by 2-wide' suggests double scanning for 160x100 in both horizontal and vertical. The low-res mode is in fact 80x25 using drawing characters, which has to be clocked for 640x200. That means a PEL in this case is actually 4 clocks by 2 clocks, not 2 clocks by 2 clocks.

Later on we get the 320x200 description which describes the PELs for that mode being 1 by 1, which going with the definition above would suggest that the PELs in this mode are 2 clocks by 1 clock.

But in the 640x200 description, the PELs are also described as being 1 by 1, and we know that corresponds to 1 clock by 1 clock.

So, the definition of PEL is not coupled with the pixel clock at all, but is rather solely a description of what the programmer can control on screen. So there is an ambiguity of whether the CGA card clocks twice for each pixel when in the 320x240 mode.

So I turned to the [6845 Data Sheet](https://github.com/86Box/86Box/files/11438399/6845.pdf). The data sheet shows that 1:1 mapping of horizontal pixels to the input clock (but this is just a recommendation). Sheet four of the schematic of the CGA card will show two separate clocks fed to the 6845 depending on whether 40x25/320x200 or 80x25/640x200/160x200 mode is used. So the 6845 is being clocked by two separate timings depending on the mode. On the same sheet, you'll also find that the pixel dot clock that feeds the shift register has the same setup. The dot clock uses 14 MHz for 80x25 and 7 MHz for 40x25. So the card clocks a different number of cycles to control the CRTC and it clocks a different number of pixels for actually drawing the lines directed by the CRTC depending on the graphics mode.

**Reply by OBattler** [reply]
I think the answer is in simple maths: the period is the inverse of the frequency, so if the frequency halves, the period doubles. Which means that in each pixel's state, the beam remains for twice as long as it normally would. Since the grid on the monitor remains the same, that gives the perception of horizontal doubling.

**Reply by anikom15** [reply]
Sure, but as far as what the CGA card is doing, it is clocking half the pixels. It's not clocking the same pixel twice. Having the driver fill out the bitmap to whatever the maximum capable horizontal resolution of the card is is not consistent with the convention used by other emulators, and isn't even consistent within 86Box itself. For example, the SVGA cards don't fill out a 640 bitmap to a larger bitmap when in 640x480 mode, even though various SVGA cards are capable of higher resolutions.

I don't see a drawback to having the bitmap correspond to the dot clock, and it has the benefit of being what third-party shaders expect. To make certain I implemented it correctly, I tested my changes with CGA_COMP and the behavior is identical, so this shouldn't break compatibility.

**Reply by OBattler** [reply]
From what I know, the various (S)VGA cards have RAM DAC's (and/or clock chips) which support more than just the default two clock speeds, precisely for the other modes. And the monitor had to support a particular resolution to be able to display it. So drawing 640x480 stretched to, say, 800x600, wouldn't be correct.

On the other hand, CGA only has one vertical resolution, with the screen always displaying the same horizontal resolution as well, just that the effective resolution is halved in the half-clock-rate mode, which results in it looking doubled in practice. So horizontally, our output is correct. Vertically, it's not and I agree that should be rectified. Also, if we were to output in 320x200, then use the renderer to stretch to 640x400, then, unless the worst resize method were used, it would appear horizontally blurrier than how it appears on the real monitor.

> I tested my changes with CGA_COMP and the behavior is identical, so this shouldn't break compatibility.

Have you tested 8088 MPH as well?

**Reply by OBattler** [reply]
Also:
> it is clocking half the pixels. It's not clocking the same pixel twice.

It is clocking half the pixels and the duration of it is that expected of all the pixels, so each pixel is clocked for twice the time.

To illustrate what happens:
Full clock speed, 01010101 pattern:
```
 ┌─┐ ┌─┐ ┌─┐ ┌─
─┘ └─┘ └─┘ └─┘
```
Full clock speed, 00110011 pattern:
```
   ┌───┐   ┌───
───┘   └───┘
```
Half clock speed, 0101 pattern:
```
   ┌───┐   ┌───
───┘   └───┘
```
In practice, the last two yield the same result.

**Reply by OBattler** [reply]
Now, what we should do, is making the timings are correct (I don't think we lower the clock rate in lower resolutions).

**Reply by OBattler** [reply]
Now that I think of it - actually, there should be a difference between the bottom two - the half clock speed pattern should have blurrier edges since it takes twice as long for the beam's state to change as well. To emulate that accurately, it should be determined how long it takes for the beam to transition from completely off to on at full strength, and then divide accordingly.

**Reply by anikom15** [reply]
I was able to get my changes to pass most of 8088 MPH when compared to unchanged 86Box itself. There are a couple of issues that need to be worked out though. The vertical scolling part with the flying car and the IBM logo scenes don't render across the whole screen.

**Reply by OBattler** [reply]
Also, we have a good reason to render CGA at a fixed resolution - Area 5150. Even the current emulation resizes too much for it, let alone if we actually rendered 320x200 as 320x200.

**Reply by anikom15** [reply]
I think I will separate vertical resolution fix into a separate branch, so that can get pulled. The vertical change should easily pass both 8088 MPH and Area 5150.

**Comment by OBattler** [comment]
And there's also a very good reason why the card does it - the composite output must be 480i, ie. standard NTSC. So the real card has to do any doubling necessary.

**Reply by dbalsom** [reply]
The card doesn't do any vertical doubling. There's no circuitry for it. It emits 262 scanlines per frame at 59.93 Hz.  A broadcast 480i signal would have two different fields increasing the effective vertical resolution but at a frame rate of ~30 fps. The CGA gives up spatial for temporal resolution.  A TV treats two successive frames output by the CGA as two fields and displays them one after the other.

**Comment by dbalsom** [comment]
The CRTC on the CGA card can be configured to output in excess of 640 vertical lines. 
For example, this effect from Area 5150 is effectively 768x246.  
![image](https://github.com/86Box/86Box/assets/7229541/edcb268f-a2ff-4e70-8465-420b5e1fadd4)

I recommend rendering into a fixed NTSC-like field of 912x262 representing 238944 CGA clock ticks per frame, and displaying some rectangular viewport into that. In 40 column mode you could draw every pixel twice, or change your field to 456x262 representing the half pixel clock in that mode.

**Reply by OBattler** [reply]
I agree with that idea.
